File: src/semp_workflow/modules/q_sub.py
"""Queue Subscription module - q_sub.add, q_sub.delete."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient, ALREADY_EXISTS
from ..semp.helpers import enc
from .base import BaseModule

logger = logging.getLogger(__name__)


class SubscriptionAdd(BaseModule):
    description = "Add a topic subscription to a queue. Skipped if the subscription already exists."
    params = {
        "queueName":         {"type": "string", "required": True, "description": "Name of the queue to subscribe"},
        "subscriptionTopic": {"type": "string", "required": True, "description": "Topic string to subscribe to (wildcards supported, e.g. FCM/SAP/>)"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        queue_name = args.get("queueName", "")
        topic = args.get("subscriptionTopic", "")
        if not queue_name or not topic:
            return ActionResult(
                ResultStatus.FAILED, "Missing required args: queueName, subscriptionTopic"
            )

        if dry_run:
            path = f"queues/{enc(queue_name)}/subscriptions/{enc(topic)}"
            try:
                exists, _ = client.exists(path)
            except SEMPError:
                exists = False
            if exists:
                return ActionResult(ResultStatus.SKIPPED, f"Subscription already exists")
            return ActionResult(ResultStatus.DRYRUN, f"Would add subscription '{topic}' to queue '{queue_name}'")

        try:
            payload = {
                "queueName": queue_name,
                "subscriptionTopic": topic,
            }
            client.create(f"queues/{enc(queue_name)}/subscriptions", payload)
            return ActionResult(
                ResultStatus.OK,
                f"Subscription '{topic}' added to queue '{queue_name}'",            )
        except SEMPError as e:
            if e.semp_code == ALREADY_EXISTS:
                return ActionResult(
                    ResultStatus.SKIPPED,
                    f"Subscription '{topic}' already exists on queue '{queue_name}'",
                )
            return ActionResult(
                ResultStatus.FAILED,
                f"Failed to add subscription: {e}",
            )


class SubscriptionDelete(BaseModule):
    description = "Remove a topic subscription from a queue. Skipped if the subscription does not exist."
    params = {
        "queueName":         {"type": "string", "required": True, "description": "Name of the queue"},
        "subscriptionTopic": {"type": "string", "required": True, "description": "Topic string to unsubscribe"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        queue_name = args.get("queueName", "")
        topic = args.get("subscriptionTopic", "")
        if not queue_name or not topic:
            return ActionResult(
                ResultStatus.FAILED, "Missing required args: queueName, subscriptionTopic"
            )

        path = f"queues/{enc(queue_name)}/subscriptions/{enc(topic)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking subscription: {e}")

        if not exists:
            return ActionResult(
                ResultStatus.SKIPPED,
                f"Subscription '{topic}' does not exist on queue '{queue_name}'",
            )

        if dry_run:
            return ActionResult(
                ResultStatus.DRYRUN,
                f"Would remove subscription '{topic}' from queue '{queue_name}'",
            )

        try:
            client.delete(path)
            return ActionResult(
                ResultStatus.OK,
                f"Subscription '{topic}' removed from queue '{queue_name}'",            )
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Failed to remove subscription: {e}")


MODULES = {
    "q_sub.add": SubscriptionAdd,
    "q_sub.delete": SubscriptionDelete,
}

File: internal/modules/q_sub.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("q_sub.add", subscriptionAdd{})
	register("q_sub.delete", subscriptionDelete{})
}

type subscriptionAdd struct{}

func (subscriptionAdd) Description() string {
	return "Add a topic subscription to a queue. Skipped if the subscription already exists."
}

func (subscriptionAdd) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "queueName", Type: "string", Required: true, Description: "Name of the queue to subscribe"},
		{Name: "subscriptionTopic", Type: "string", Required: true, Description: "Topic string to subscribe to (wildcards supported, e.g. SITEA/SAP/>)"},
	}
}

func (subscriptionAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	queueName := argStr(args, "queueName", "")
	topic := argStr(args, "subscriptionTopic", "")
	if queueName == "" || topic == "" {
		return failed("Missing required args: queueName, subscriptionTopic")
	}

	if dryRun {
		path := fmt.Sprintf("queues/%s/subscriptions/%s", semp.Enc(queueName), semp.Enc(topic))
		exists, _, err := c.Exists(path)
		if err != nil {
			exists = false
		}
		if exists {
			return skipped("Subscription already exists")
		}
		return dryrun(fmt.Sprintf("Would add subscription '%s' to queue '%s'", topic, queueName))
	}

	payload := map[string]any{
		"queueName":         queueName,
		"subscriptionTopic": topic,
	}
	if _, err := c.Create(fmt.Sprintf("queues/%s/subscriptions", semp.Enc(queueName)), payload); err != nil {
		if sempCodeOf(err) == semp.AlreadyExists {
			return skipped(fmt.Sprintf("Subscription '%s' already exists on queue '%s'", topic, queueName))
		}
		return failed(fmt.Sprintf("Failed to add subscription: %s", err))
	}
	return ok(fmt.Sprintf("Subscription '%s' added to queue '%s'", topic, queueName))
}

type subscriptionDelete struct{}

func (subscriptionDelete) Description() string {
	return "Remove a topic subscription from a queue. Skipped if the subscription does not exist."
}

func (subscriptionDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "queueName", Type: "string", Required: true, Description: "Name of the queue"},
		{Name: "subscriptionTopic", Type: "string", Required: true, Description: "Topic string to unsubscribe"},
	}
}

func (subscriptionDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	queueName := argStr(args, "queueName", "")
	topic := argStr(args, "subscriptionTopic", "")
	if queueName == "" || topic == "" {
		return failed("Missing required args: queueName, subscriptionTopic")
	}

	path := fmt.Sprintf("queues/%s/subscriptions/%s", semp.Enc(queueName), semp.Enc(topic))

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking subscription: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("Subscription '%s' does not exist on queue '%s'", topic, queueName))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would remove subscription '%s' from queue '%s'", topic, queueName))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to remove subscription: %s", err))
	}
	return ok(fmt.Sprintf("Subscription '%s' removed from queue '%s'", topic, queueName))
}
